### DISCUSSION TITLE
First release

### DISCUSSION THREAD

**Discussion by automartin5000** [projen/projen#3205] [Q&A]
I'm working on my first Projen project and I'm a bit confused on the way releases are supposed to work. I've read [this](https://projen.io/releases.html) a couple times, but still don't quite follow how/when git tags are applied.

For context, I don't currently have access to GitHub Actions, so I've been iterating locally on everything. My development process has been `delete existing version 0.0.0.0 from artifact repo` -> `projen release` -> `npm publish`. However I'm now to the point where I want to cut an actual release and start a new branch. When I run `git tag` in the repo, I noticed no git tags even though I've run `projen release` a number of times and every package being created still results in a `0.0.0.0` version. I figured I could try `projen bump`, but it sounds like that shouldn't be required? I'm also wondering if a merge to `main` is required for the tag to be applied?

**Comment by mrgrain** [comment]
Did you set this?

```ts
releaseTrigger: ReleaseTrigger.manual(),
```

**Reply by automartin5000** [reply]
I didn't. What does that do? Tag on `projen release`?

I noticed `ReleaseTrigger` has 3 methods. I can imagine what `scheduled()` does, but what is `continuous()`?

**Reply by mrgrain** [reply]
> I didn't. What does that do? Tag on `projen release`?
> 
> I noticed `ReleaseTrigger` has 3 methods. I can imagine what `scheduled()` does, but what is `continuous()`?

`continous()` releases every eligible commit. 

You'll need to set it to manual. The other two require GitHub Actions to work.

**Reply by automartin5000** [reply]
Ok, so I run release and the tag appears, but a couple things are unexpected.
First, I created a new `feat` commit with README changes to try to trigger a `0.1` release, but instead when I ran `PRERELEASE=rc projen release` I got a new git tag (yay), but the version is `v0.0.1-rc.0` 😢 . Also, the packages in `dist` still stay `0.0.0`. I tried `projen package`, but that got no different results.

**Reply by automartin5000** [reply]
Sorry, I ran into another thing. I created a new branch. This one should ideally ideally be version `0.2.0` (the above branch was supposed to be `0.1.0`). Running `PRERELEASE=alpha projen release` results in the following:
```
 * [new tag]         v0.0.0 -> v0.0.0
 * [new tag]         v0.0.1-rc.0 -> v0.0.1-rc.0
 ! [remote rejected] main -> main (protected branch hook declined)
```
I'm not on `main` so I'm not sure why it's trying to tag that branch. But running the exact same command immediately after is successful and packages everything correctly, albeit as version `0.0.0` even though the tag is `v0.0.1-alpha.0`.

**Reply by mrgrain** [reply]
> Ok, so I run release and the tag appears, but a couple things are unexpected.
First, I created a new feat commit with README changes to try to trigger a 0.1 release, but instead when I ran PRERELEASE=rc projen release I got a new git tag (yay), but the version is v0.0.1-rc.0 😢

Works as expected. In major version zero (0.x.y) releases, a `feat` does not cause the `minor` version to change according to semver.

> Also, the packages in dist still stay 0.0.0. I tried projen package, but that got no different results. 

I can't reproduce this, the dist package contains the correct version for me. See screenshot.
<img width="491" alt="image" src="https://github.com/projen/projen/assets/379814/3cc2e680-c8b7-41b6-b068-c1ca27c1849f">

**Reply by mrgrain** [reply]
> Running `PRERELEASE=alpha projen release` results in the following:
> I'm not on `main` so I'm not sure why it's trying to tag that branch. But running the exact same command immediately after is successful and packages everything correctly, albeit as version `0.0.0` even though the tag is `v0.0.1-alpha.0`.

Your attempt to actually publish the tag to the world has failed because you appear to be lacking permissions to push tags to your `main` branch. Running the command twice will work because the tag is found locally and the release command determines that there is nothing to do. This is also why the version in the dist package is `0.0.0` - a bump did not occur.

---

If you need more control over specific versions, you might want to look into the `releaseBranches` feature.

**Reply by automartin5000** [reply]
> Your attempt to actually publish the tag to the world has failed because you appear to be lacking permissions to push tags to your `main` branch. 

Yeah we have branch protections set up so that was expected. I was just confused why it was attempting to tag `main`?

> Running the command twice will work because the tag is found locally and the release command determines that there is nothing to do. This is also why the version in the dist package is `0.0.0` - a bump did not occur.
> 
> If you need more control over specific versions, you might want to look into the `releaseBranches` feature.

Got it, thanks! I'll look into that

**Reply by automartin5000** [reply]
> Works as expected. In major version zero (0.x.y) releases, a feat does not cause the minor version to change according to semver.

Is that a Projen specific implementation? Not seeing that in semver [documentation](https://semver.org/#spec-item-4) 🤔

**Reply by mrgrain** [reply]
[No, it's semver, rule 4](https://semver.org/#spec-item-4). Granted the leap to versions is not obvious. It's due to how [npm handles it](https://stackoverflow.com/a/22345808/3778831) and the connection to what effect `feat` vs `fix` have is coming from [conventional-changelog](https://github.com/conventional-changelog), specifically the `standard-version` tool that projen is using internally.

**Reply by mrgrain** [reply]
> Yeah we have branch protections set up so that was expected. I was just confused why it was attempting to tag main?

It doesn't for me, so I'm as confused as you are. Saying that, the output you posted isn't from tagging, but from pushing. Also I was wrong, it doesn't just tag. It also creates a changelog commit and then attempts to push that.

**Reply by automartin5000** [reply]
Ok, I think I have the manual build/release/publish process down. Thanks!
